File: packages/sdk-ts/src/lib/webhooks.ts
/**
 * Webhook signature verification (spec 04 §2.4): `Webhook-Signature: t=<unix>,v1=<hex hmac-sha256(secret, t + "." + body)>`.
 * Web Crypto only, so it runs in Node 18+, Deno, Bun and Workers.
 */
const encoder = new TextEncoder();

async function hmacHex(secret: string, message: string): Promise<string> {
  const key = await crypto.subtle.importKey('raw', encoder.encode(secret), { name: 'HMAC', hash: 'SHA-256' }, false, ['sign']);
  const sig = await crypto.subtle.sign('HMAC', key, encoder.encode(message));
  return [...new Uint8Array(sig)].map((b) => b.toString(16).padStart(2, '0')).join('');
}

export interface VerifyOptions {
  /** Maximum age of the timestamp in seconds (default 300). */
  toleranceSeconds?: number;
  /** Unix seconds "now"; tests pass a fixed clock. */
  now?: number;
}

/**
 * Returns true when the header matches the raw body. Pass the body exactly as received (not
 * re-serialised JSON): the signature covers the bytes on the wire.
 */
export async function verifyWebhookSignature(
  secret: string,
  signatureHeader: string | null | undefined,
  rawBody: string,
  options: VerifyOptions = {},
): Promise<boolean> {
  if (!signatureHeader || !secret) return false;
  const parts = new Map(signatureHeader.split(',').map((kv) => kv.trim().split('=') as [string, string]));
  const t = Number(parts.get('t'));
  const v1 = (parts.get('v1') ?? '').toLowerCase();
  if (!Number.isFinite(t) || !/^[0-9a-f]{64}$/.test(v1)) return false;
  const now = options.now ?? Math.floor(Date.now() / 1000);
  if (Math.abs(now - t) > (options.toleranceSeconds ?? 300)) return false;
  const expected = await hmacHex(secret, `${t}.${rawBody}`);
  let diff = 0;
  for (let i = 0; i < expected.length; i++) diff |= expected.charCodeAt(i) ^ v1.charCodeAt(i);
  return diff === 0;
}

/** Parses a verified event body. Throws when the signature does not match. */
export async function parseWebhookEvent<T = { id: string; type: string; created_at: string; workspace_id: string | null; data: unknown }>(
  secret: string,
  signatureHeader: string | null | undefined,
  rawBody: string,
  options: VerifyOptions = {},
): Promise<T> {
  if (!(await verifyWebhookSignature(secret, signatureHeader, rawBody, options)))
    throw new Error('invalid webhook signature');
  return JSON.parse(rawBody) as T;
}

File: packages/sdk-python/src/formfeed/webhooks.py
"""Webhook signature verification (spec 04 §2.4): ``Webhook-Signature: t=<unix>,v1=<hex hmac-sha256(secret, t + "." + body)>``."""

from __future__ import annotations

import hashlib
import hmac
import json
import re
import time
from typing import Any

from .models import WebhookEvent

_HEX64 = re.compile(r"^[0-9a-f]{64}$")


def verify_webhook_signature(
    secret: str,
    signature_header: str | None,
    raw_body: bytes | str,
    *,
    tolerance_seconds: int = 300,
    now: int | None = None,
) -> bool:
    """True when the header matches the raw body. Pass the body exactly as received: the signature covers the bytes on the wire."""
    if not signature_header or not secret:
        return False
    parts: dict[str, str] = {}
    for kv in signature_header.split(","):
        key, _, value = kv.strip().partition("=")
        parts[key] = value
    try:
        t = int(parts.get("t", ""))
    except ValueError:
        return False
    v1 = parts.get("v1", "").lower()
    if not _HEX64.match(v1):
        return False
    current = now if now is not None else int(time.time())
    if abs(current - t) > tolerance_seconds:
        return False
    body = raw_body.encode("utf-8") if isinstance(raw_body, str) else raw_body
    expected = hmac.new(secret.encode("utf-8"), f"{t}.".encode("utf-8") + body, hashlib.sha256).hexdigest()
    return hmac.compare_digest(expected, v1)


def parse_webhook_event(
    secret: str,
    signature_header: str | None,
    raw_body: bytes | str,
    *,
    tolerance_seconds: int = 300,
    now: int | None = None,
) -> WebhookEvent:
    """Parses a verified event body; raises ``ValueError`` when the signature does not match."""
    if not verify_webhook_signature(secret, signature_header, raw_body, tolerance_seconds=tolerance_seconds, now=now):
        raise ValueError("invalid webhook signature")
    payload: dict[str, Any] = json.loads(raw_body)
    return WebhookEvent.model_validate(payload)
